### PR TITLE
[Coop] Convert performance counters and performance counter categories.

### PR DESCRIPTION
**Pull request by jaykrell**
This is similar old unmerged PR, it was approved by one, requested changes another.
Some of the requested changes were made, some were not, some were questionable requests, for example, it was requested to restore creating handles in loops.
Some can be reconsidered.

### CODE REVIEW

**Comment by jaykrell** [comment]
@monojenkins build deb with monolite

**Comment by jaykrell** [comment]
@monojenkins build failed

**Comment by jaykrell** [comment]
@monojenkins build failed

**Comment by jaykrell** [comment]
@monojenkins build failed

**Comment by jaykrell** [comment]
@monojenkins build failed

**Comment by jaykrell** [comment]
@monojenkins build Linux x64 Coop Suspend

**Comment by jaykrell** [comment]
@monojenkins build Windows x64

**Comment by jaykrell** [comment]
@monojenkins build Windows i386

**Comment by jaykrell** [comment]
@monojenkins build Windows x64 FullAOT

**Comment by jaykrell** [comment]
@monojenkins build Windows x64 C++

**Comment by jaykrell** [comment]
@monojenkins build Linux x64 Coop Suspend

**Comment by jaykrell** [comment]
Corlib version change is currently broken on Windows.

**Comment by jaykrell** [comment]
@monojenkins build deb with monolite

**Comment by jaykrell** [comment]
@monojenkins build failed

File: mono/metadata/mono-perfcounters.c [hunk 1]
```diff
@@ -1591,98 +1603,123 @@ mono_perfcounter_instance_exists (MonoString *instance, MonoString *category)
...
 		goto leave;
 	}
 
 	for (i = 0; i < NUM_CATEGORIES; ++i) {
 		const CategoryDesc *cdesc = &predef_categories [i];
-		MonoString *name = mono_string_new_checked (domain, cdesc->name, error);
+		MonoStringHandle name;
+		SETUP_ICALL_FRAME;
+		name = mono_string_new_utf8_len (domain, cdesc->name, cdesc->name_length, error);
+		goto_if_nok (error, clear_icall_frame1);
```
Comment (jaykrell) [inline mono/metadata/mono-perfcounters.c]: The `goto` carefully chosen label is a little ugly.
Alternatives:
 `_assign` which has other optimizations but was rejected
put success path under `if success` and so error path falls to the label.

File: mono/metadata/mono-perfcounters.c [hunk 2]
```diff
@@ -1693,40 +1730,48 @@ get_string_array (void **array, int count, gboolean is_process, MonoError *error
...
 		}
-		MonoString *str = mono_string_new_checked (domain, p, error);
+		MonoStringHandle str;
+		SETUP_ICALL_FRAME;
+		str = mono_string_new_utf8_len (domain, p, strlen (p), error);
 		if (p != buf)
 			g_free (p);
-		return_val_if_nok (error, NULL);
-		mono_array_setref_internal (res, i, str);
+		MONO_HANDLE_ARRAY_SETREF (res, i, str);
```
Comment (jaykrell) [inline mono/metadata/mono-perfcounters.c]: Previous code skipped the array_setref in the error path.

Comment (jaykrell) [reply mono/metadata/mono-perfcounters.c]: But the array is probably dead/unseen in that case, just garbage, in which case ok. Same with next chunk.

**Comment by jaykrell** [comment]
@monojenkins build failed

File: mono/metadata/mono-perfcounters.c [hunk 3]
```diff
@@ -1534,52 +1552,46 @@ mono_perfcounter_create (MonoString *category, MonoString *help, int type, MonoA
 	cat->counters_data_size = counters_data_size;
 	/* now copy the vaiable data */
 	p = cat->name;
-	strcpy (p, name);
-	p += strlen (name) + 1;
-	strcpy (p, chelp);
-	p += strlen (chelp) + 1;
+	p = append_str (p, name);
```
Comment (jaykrell) [inline mono/metadata/mono-perfcounters.c]: Diff getting a bit noisy/superfluous. Could be reduced.

**Comment by jaykrell** [comment]
@monojenkins build failed

**Comment by jaykrell** [comment]
@monojenkins build deb with monolite

**Comment by jaykrell** [comment]
@monojenkins build failed

**Comment by jaykrell** [comment]
@monojenkins build deb with monolite

**Comment by jaykrell** [comment]
@monojenkins build failed

File: mono/metadata/mono-perfcounters.c [hunk 4]
```diff
@@ -1591,98 +1605,121 @@ mono_perfcounter_instance_exists (MonoString *instance, MonoString *category)
...
 		goto_if_nok (error, leave);
-		mono_array_setref_internal (res, i, name);
 	}
 	for (tmp = custom_categories; tmp; tmp = tmp->next) {
 		SharedCategory *scat = (SharedCategory *)tmp->data;
-		MonoString *name = mono_string_new_checked (domain, scat->name, error);
+		MonoStringHandle name;
+		SETUP_ICALL_FRAME;
+		name = mono_string_new_utf8_len (domain, scat->name, strlen (scat->name), error);
+		if (is_ok (error))
```
Comment (jaykrell) [inline mono/metadata/mono-perfcounters.c]: `if` is debatable here, but I err toward preserving the semantics of the original.

File: mono/metadata/mono-perfcounters.c [hunk 5]
```diff
@@ -1591,98 +1605,121 @@ mono_perfcounter_instance_exists (MonoString *instance, MonoString *category)
...
 		for (i = cdesc->first_counter; i < cdesc [1].first_counter; ++i) {
 			const CounterDesc *desc = &predef_counters [i];
-			MonoString *name = mono_string_new_checked (domain, desc->name, error);
-			if (mono_error_set_pending_exception (error))
-				return NULL;
-			mono_array_setref_internal (res, i - cdesc->first_counter, name);
+			MonoStringHandle name;
+			SETUP_ICALL_FRAME;
+			name = mono_string_new_utf8_len (domain, desc->name, desc->name_length, error);
+			if (is_ok (error))
```
Comment (jaykrell) [inline mono/metadata/mono-perfcounters.c]: `if` is debatable here, but I err toward preserving the semantics of the original.

File: mono/metadata/mono-perfcounters.c [hunk 6]
```diff
@@ -1591,98 +1605,121 @@ mono_perfcounter_instance_exists (MonoString *instance, MonoString *category)
...
+			res = NULL_HANDLE_ARRAY;
+			goto leave;
 		}
 
 		for (i = 0; i < scat->num_counters; ++i) {
-			MonoString *str = mono_string_new_checked (domain, counter->name, error);
+			MonoStringHandle str;
+			SETUP_ICALL_FRAME;
+			str = mono_string_new_utf8_len (domain, counter->name, strlen (counter->name), error);
+			if (is_ok (error))
```
Comment (jaykrell) [inline mono/metadata/mono-perfcounters.c]: `if` is debatable here, but I err toward preserving the semantics of the original.

**Comment by lambdageek** [comment]
This seems identical to https://github.com/mono/mono/pull/12276 with none of the issues that were raised there being addressed.

**Comment by jaykrell** [comment]
You had approved it.
Several issues addressed. Not all.
Some of them reflected a lack of understanding, of avoiding creating handles in loops.
I can address some more. But some of it is deliberate. Just because someone raises an issue does not make it correct to.

**Comment by jaykrell** [comment]
Most of the former PR feedback has been addressed by restoring the older larger code. Not all of it.

**Comment by jaykrell** [comment]
@monojenkins build deb with monolite

**Comment by jaykrell** [comment]
Better?

**Comment by jaykrell** [comment]
@monojenkins build failed

**Comment by jaykrell** [comment]
Better now?

**Comment by NattyNarwhal** [comment]
making a note that this may have caused a [regression](https://gist.github.com/NattyNarwhal/6f516d444981fa8c698d57a29ef40c29) [on aix/ppc64](https://jenkins.mono-project.com/job/test-mono-mainline-community/label=aix-ppc64/4576/), or it's exposed a possible PPC JIT bug I'm not 100% sure of the cause (I can't verify, the ubuntu/ppc64be system I have access to is a ppc32 userland and that confused mono at build time, and CI for ppc64le dies somewhere early in the corlib tests instead) but I have seen before - noticed because it causes tests to crash if they hit it.

if it is a PPC JIT bug, my theory is the number of args is greater than the number of PPC GPRs for args, and that's an issue; so the older style call was fine due to lower number of args, but i need to look into that

**Comment by jaykrell** [comment]
There are possible workarounds.
Is the problem specific to icalls or not? I am more sympathetic if it is icall-specific because that is a closed set. We can pass a byref struct of params, or call strlen on native side, etc. or remove this feature from ppc.

C# to C# calls should support fairly arbitrary parameter counts and types though. And with p/invoke.

**Comment by vargaz** [comment]
@NattyNarwhal what is the failure ?

**Comment by jaykrell** [comment]
We can attempt workaround, at least to collect data.
https://github.com/mono/mono/pull/14153

**Comment by jaykrell** [comment]
@NattyNarwhal please trigger a PowerPC build on https://github.com/mono/mono/pull/14153 after the others succeed.

**Pull request merged**